File: src/server/map-generate-progress.ts
export interface MapGenerateProgressSnapshot {
  completedSteps: number;
  totalSteps: number;
  phase: 'adt' | 'convert';
  taskName?: string;
  tileIndex?: number;
  tileCount?: number;
  stepsPerTile?: number;
  currentTile?: { x: number; y: number };
  creatureCompleted?: number;
  creatureTotal?: number;
}

const progressStore = new Map<string, MapGenerateProgressSnapshot>();

export function initMapGenerateProgress(
  key: string,
  params: {
    adtTotalSteps: number;
    convertSteps: number;
    tileCount: number;
    stepsPerTile: number;
  },
): void {
  progressStore.set(key, {
    completedSteps: 0,
    totalSteps: params.adtTotalSteps + params.convertSteps,
    phase: 'adt',
    taskName: 'Exporting tiles',
    tileIndex: 0,
    tileCount: params.tileCount,
    stepsPerTile: params.stepsPerTile,
  });
}

export function updateMapGenerateTotalSteps(key: string, convertSteps: number, adtTotalSteps: number): void {
  const snap = progressStore.get(key);
  if (!snap) return;
  snap.totalSteps = adtTotalSteps + convertSteps;
}

export function setMapGeneratePhase(
  key: string,
  phase: 'adt' | 'convert',
  taskName: string,
): void {
  const snap = progressStore.get(key);
  if (!snap) return;
  snap.phase = phase;
  snap.taskName = taskName;
  if (phase === 'convert') {
    snap.tileIndex = undefined;
    snap.currentTile = undefined;
  }
}

export function syncAdtProgress(
  key: string,
  adtSnap: {
    completedSteps: number;
    tileIndex: number;
    tileCount: number;
    stepsPerTile: number;
    currentTile?: { x: number; y: number };
    taskName?: string;
  },
): void {
  const snap = progressStore.get(key);
  if (!snap) return;
  snap.phase = 'adt';
  snap.completedSteps = adtSnap.completedSteps;
  snap.tileIndex = adtSnap.tileIndex;
  snap.tileCount = adtSnap.tileCount;
  snap.stepsPerTile = adtSnap.stepsPerTile;
  snap.currentTile = adtSnap.currentTile;
  snap.taskName = adtSnap.taskName ?? 'Exporting tiles';
}

export function advanceMapGenerateProgress(
  key: string,
  adtTotalSteps: number,
  convertCompletedSteps: number,
  taskName: string,
  creatureProgress?: { completed: number; total: number },
): void {
  const snap = progressStore.get(key);
  if (!snap) return;
  snap.phase = 'convert';
  snap.completedSteps = adtTotalSteps + convertCompletedSteps;
  snap.taskName = taskName;
  snap.tileIndex = undefined;
  snap.currentTile = undefined;
  if (creatureProgress) {
    snap.creatureCompleted = creatureProgress.completed;
    snap.creatureTotal = creatureProgress.total;
  } else {
    snap.creatureCompleted = undefined;
    snap.creatureTotal = undefined;
  }
}

export function getMapGenerateProgress(key: string): MapGenerateProgressSnapshot | undefined {
  return progressStore.get(key);
}

export function clearMapGenerateProgress(key: string): void {
  progressStore.delete(key);
}

export function toProgressPercent(snap: MapGenerateProgressSnapshot): number {
  return snap.totalSteps > 0
    ? Math.min(100, Math.round((snap.completedSteps / snap.totalSteps) * 100))
    : 0;
}

File: golang/internal/server/api/map_generate_progress.go
package api

import (
	"sync"
)

type mapGeneratePhase string

const (
	phaseADT     mapGeneratePhase = "adt"
	phaseConvert mapGeneratePhase = "convert"
)

type mapGenerateSnapshot struct {
	CompletedSteps   int
	TotalSteps       int
	Phase            mapGeneratePhase
	TaskName         string
	TileIndex        *int
	TileCount        *int
	StepsPerTile     *int
	CurrentTile      *tileCoord
	CreatureCompleted *int
	CreatureTotal    *int
}

type tileCoord struct {
	X int `json:"x"`
	Y int `json:"y"`
}

var (
	mapGenerateMu    sync.Mutex
	mapGenerateStore = map[string]*mapGenerateSnapshot{}
)

func initMapGenerateProgress(key string, adtTotalSteps, convertSteps, tileCount, stepsPerTile int) {
	mapGenerateMu.Lock()
	defer mapGenerateMu.Unlock()
	ti := 0
	tc := tileCount
	spt := stepsPerTile
	mapGenerateStore[key] = &mapGenerateSnapshot{
		CompletedSteps: 0,
		TotalSteps:     adtTotalSteps + convertSteps,
		Phase:          phaseADT,
		TaskName:       "Exporting tiles",
		TileIndex:      &ti,
		TileCount:      &tc,
		StepsPerTile:   &spt,
	}
}

func updateMapGenerateTotalSteps(key string, convertSteps, adtTotalSteps int) {
	mapGenerateMu.Lock()
	defer mapGenerateMu.Unlock()
	if snap, ok := mapGenerateStore[key]; ok {
		snap.TotalSteps = adtTotalSteps + convertSteps
	}
}

func setMapGeneratePhase(key string, phase mapGeneratePhase, taskName string) {
	mapGenerateMu.Lock()
	defer mapGenerateMu.Unlock()
	if snap, ok := mapGenerateStore[key]; ok {
		snap.Phase = phase
		snap.TaskName = taskName
		if phase == phaseConvert {
			snap.TileIndex = nil
			snap.CurrentTile = nil
		}
	}
}

func syncAdtProgress(key string, completedSteps, tileIndex, tileCount, stepsPerTile int, currentTile *tileCoord, taskName string) {
	mapGenerateMu.Lock()
	defer mapGenerateMu.Unlock()
	snap, ok := mapGenerateStore[key]
	if !ok {
		return
	}
	snap.Phase = phaseADT
	snap.CompletedSteps = completedSteps
	ti := tileIndex
	tc := tileCount
	spt := stepsPerTile
	snap.TileIndex = &ti
	snap.TileCount = &tc
	snap.StepsPerTile = &spt
	snap.CurrentTile = currentTile
	if taskName != "" {
		snap.TaskName = taskName
	} else {
		snap.TaskName = "Exporting tiles"
	}
}

func advanceMapGenerateProgress(key string, adtTotalSteps, convertCompleted int, taskName string, creatureCompleted, creatureTotal *int) {
	mapGenerateMu.Lock()
	defer mapGenerateMu.Unlock()
	snap, ok := mapGenerateStore[key]
	if !ok {
		return
	}
	snap.CompletedSteps = adtTotalSteps + convertCompleted
	snap.TaskName = taskName
	snap.CreatureCompleted = creatureCompleted
	snap.CreatureTotal = creatureTotal
}

func getMapGenerateProgress(key string) *mapGenerateSnapshot {
	mapGenerateMu.Lock()
	defer mapGenerateMu.Unlock()
	return mapGenerateStore[key]
}

func clearMapGenerateProgress(key string) {
	mapGenerateMu.Lock()
	defer mapGenerateMu.Unlock()
	delete(mapGenerateStore, key)
}

func toProgressPercent(snap *mapGenerateSnapshot) int {
	if snap == nil || snap.TotalSteps <= 0 {
		return 0
	}
	pct := (snap.CompletedSteps * 100) / snap.TotalSteps
	if pct > 100 {
		return 100
	}
	return pct
}
